src/linearAlgebra/BlasLapackKernels.cpp

```diff
@@ -10,6 +10,79 @@ namespace dftefe
   {
     namespace blasLapack
     {
+      namespace blasLapackKernelsInternal
+      {
+        template <typename T>
+        T
+        conjugate(const T &x)
+        {
+          return std::conj(x);
+        }
+
+        template <>
+        double
+        conjugate(const double &x)
+        {
+          return x;
+        }
+
+        template <>
+        float
+        conjugate(const float &x)
+        {
+          return x;
+        }
+
+        template <typename T1, typename T2, ScalarOp op1, ScalarOp op2>
+        class ScalarProduct
+        {
+        public:
+          static scalar_type<T1, T2>
+          prod(T1 t1, T2 t2)
+          {
+            return ((scalar_type<T1, T2>)t1) * ((scalar_type<T1, T2>)t2);
+          }
+        };
+
+        template <typename T1, typename T2>
+        class ScalarProduct<T1, T2, ScalarOp::Identity, ScalarOp::Conj>
+        {
+        public:
+          static scalar_type<T1, T2>
+          prod(T1 t1, T2 t2)
+          {
+            return ((scalar_type<T1, T2>)t1) *
+                   ((scalar_type<T1, T2>)conjugate(t2));
+          }
+        };
+
+        template <typename T1, typename T2>
+        class ScalarProduct<T1, T2, ScalarOp::Conj, ScalarOp::Identity>
+        {
+        public:
+          static scalar_type<T1, T2>
+          prod(T1 t1, T2 t2)
+          {
+            return ((scalar_type<T1, T2>)conjugate(t1)) *
+                   ((scalar_type<T1, T2>)t2);
+          }
+        };
+
+        template <typename T1, typename T2>
+        class ScalarProduct<T1, T2, ScalarOp::Conj, ScalarOp::Conj>
+        {
+        public:
+          static scalar_type<T1, T2>
+          prod(T1 t1, T2 t2)
+          {
+            return ((scalar_type<T1, T2>)conjugate(t1)) *
+                   ((scalar_type<T1, T2>)conjugate(t2));
+          }
+        };
```

Comment:
The above is a good idea, but because the functions are not inline I think the compiler may not be able to optimize the for loop. I am ok with this for now though.

Reply:
I've added inlines to all the functions. Do you think the compiler will pick it up or there is a risk of the compiler skipping it?

Reply:
I think with inline the compiler should be able to pick it up, as the inline keyword  explicitly replaces the function with the definition wherever it is called. Could you please also add the inline keyword for the `conj' function as well?